### DISCUSSION TITLE
Idiomatic way to generate custom errors

### DISCUSSION THREAD

**Discussion by skittishdev** [pest-parser/pest#931] [Q&A] [answered]
Hi,

I'm curious what the idiomatic way of converting `pest::error::Error<R>` into a custom error type is.

My crates are organized such that I have `MyError` as a custom type defined in a common crate (C1), and my parser using pest is defined in a totally separate crate (C2) in the same workspace. This is something I don't want to move away from, because the common error type will be used in other crates which will be added to my workspace that have nothing to do with the parser.

Here is what I'm doing right now.

```
#[derive(Parser)]
#[grammar = "grammars/my_grammar.pest"]
pub struct MyParser {}

impl MyParser {
    pub fn generate_ast(&self, input: &str) -> Result<TypedAST, MyError> {
        match MyParser::parse(Rule::main_program, input) {
            Ok(_pairs) => {
                Ok(TypedAST::MyType(AST::Program))
            },
            Err(err) => {
                Err(Self::convert_error(err))
            }
        }
    }

    fn convert_error<R>(value: Error<R>) -> MyError where R: RuleType {
        let (line_num, col_num) = match value.line_col {
            LineColLocation::Pos(x) => {
                (x.0, x.1)
            },
            LineColLocation::Span(start, end) => {
                (start.0, end.0)
            }
        };
        MyError::ParseError {
            line: value.line().to_string(),
            line_num,
            col_num,
            error_message: value.to_string(),
        }
    }
}
```

What I already tried:
* Defining a wrapper type (`WrapperError`) for `pest::error::Error` in C2, and trying to implement the `From` trait for `MyError` <-> `WrapperError`
  * This fails because I added the type constraint `<R: RuleType>` and it appears that `Rule`'s do not end up implementing this trait somehow? It looks like they should, but I get an error saying `From` trait bounds for `Rule` are not satisfied (even though I have it defined for `RuleType`.
```
pub(crate) struct WrapperError<R: RuleType> {
    err: Error<R>
}

impl <R: RuleType> WrapperError<R> {
    pub fn new(err: Error<R>) -> Self {
        Self { err }
    }
}

impl <R: RuleType> From<Error<R>> for WrapperError<R> {
    fn from(value: Error<R>) -> Self {
        Self { err: value }
    }
}

impl <R: RuleType> From<WrapperError<R>> for MyError {
    fn from(value: WrapperError<R>) -> Self {
        // same logic as `convert_error` above
        // I have to add the `R: RuleType` trait bound because I need access to the `line()` method.
    }
}
```

**Comment by skittishdev** [comment]
I'd like to be able to use the `?` operator is what all of this boils down to. I don't want to write these `match` statements in a bunch of places if I can help it. But then again, maybe I don't have to because as far as I can tell, once the `parse` method works, I think pest/PEG assures that my input is valid and fully parsed, and I don't have to worry about encountering a `pest::error::Error<R>` again.

**Comment by skittishdev** [comment]
@tomtau, sorry for the direct ping, but any suggestions here?

**Reply by tomtau** [reply]
It's strange. RuleType should be automatically implemented: https://github.com/pest-parser/pest/blob/857d9f61c3d03d5699338c5c625ba8a2b5bf608f/pest/src/lib.rs#L377 -- can you perhaps post the full code or the error message you're getting?

**Reply by skittishdev** [reply]
I've recreated a minimal example in this sample project [pest_errors_sample.zip](https://github.com/pest-parser/pest/files/12817394/pest_errors_sample.zip)

## When we attempt to build the project as-is, I am getting the following error:
```
 Compiling pest_parser v0.1.0
error[E0277]: `?` couldn't convert the error to `MyError`
  --> pest_parser/src/lib.rs:40:57
   |
39 | pub fn generate_ast(input: &str) -> Result<CSVAST, MyError> {
   |                                     ----------------------- expected `MyError` because of this
40 |     let mut pairs = CSVParser::parse(Rule::field, input)?;
   |                                                         ^ the trait `From<pest::error::Error<Rule>>` is not implemented for `MyError`
   |
   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
   = help: the trait `From<WrapperError<R>>` is implemented for `MyError`
   = note: required for `Result<CSVAST, MyError>` to implement `FromResidual<Result<Infallible, pest::error::Error<Rule>>>`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `pest_parser` (lib) due to previous error
```

## When  line 40 of `pest_parser/src/lib.rs` is changed to the following, it builds fine and works as expected
```
 CSVParser::parse(Rule::field, input).unwrap();
```

Output in this case:
```
 cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.03s
     Running `target/debug/pest_errors`
Generated AST for input: "-273.15": Field {
    value: -273.15,
}
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { variant: ParsingError { positives: [field], negatives: [] }, location: Pos(0), line_col: Pos((1, 1)), path: None, line: "this is not a number", continued_line: None }', pest_parser/src/lib.rs:40:58
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Reply by skittishdev** [reply]
Weird. I could have sworn the error I saw was different before (related to `RuleType` like I said in the original message). 

With the new error I'm seeing now, I guess I could write my code like this

```
pub fn generate_ast(input: &str) -> Result<CSVAST, MyError> {
    match CSVParser::parse(Rule::field, input) {
        Ok(mut pairs) => {
            let field = pairs.next().unwrap();
            Ok(CSVAST::Field{ value: str::parse(field.as_str()).unwrap()})
        },
        Err(error) => {
            Err(WrapperError::new(error))?
        }
    }
}
```

But, is there any way I can write and have it be auto-converted into `MyError`
```
let mut pairs = CSVParser::parse(Rule::field, input)?;
```

**Reply by skittishdev** [reply]
I apologize if this is more a Rust question than a pest.rs question.

In the worst case, maybe this surfaces the need/requirement to be able to generate custom errors in scenarios like this?

**Reply by tomtau** [reply]
I think you can implement `From` : https://doc.rust-lang.org/std/convert/trait.From.html which should do the conversion?

https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html#a-shortcut-for-propagating-errors-the--operator

**Reply by tomtau** [reply]
If you use [thiserror](https://docs.rs/thiserror/latest/thiserror/), you can have it generated by annotating `#[from]` in your wrapper

**Reply by skittishdev** [reply]
I can't implement the `From` trait because `MyError` is in a different crate than where the parser lives. Same for `thiserror` (I'd have to take a dependency on `pest` in my `common` crate in the above example). In fact, I am already using `thiserror` in this cargo workspace.

**Reply by skittishdev** [reply]
That's why I created the `WrapperError` in the crate which does have a dependency on `pest`, and `WrapperError` implements `From` for `pest::error::Error<R>`.

**Reply by tomtau** [reply, marked as answer]
Oh, I missed it was `MyError`, not `WrapperError`... maybe you can have a helper (private) function `generate_ast_helper` that returns `WrapperError` as the error type that's called by `generate_ast`... in that way, you can use `?` in both?

**Reply by skittishdev** [reply]
Ah, of course. This worked. Awesome! Thank you so much!

```
pub fn generate_ast(input: &str) -> Result<CSVAST, MyError> {
    Ok(generate_ast_helper(input)?)
}

fn generate_ast_helper(input: &str) -> Result<CSVAST, WrapperError<Rule>> {
    let field = CSVParser::parse(Rule::field, input)?;
    Ok(CSVAST::Field{ value: str::parse(field.as_str()).unwrap()})
}
```
